### PR TITLE
Machine uart

### PR DESCRIPTION
**Pull request by IFX-Anusha**
Machine. UART Module Implementation with Tests and Documentation.
HIL test setup yet to be updated

### CODE REVIEW

File: tests/psoc6/hw_ext/uart.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    tx_data = b"Hi this is data"
+    uart.write(tx_data)
+    time.sleep_ms(100)
+    while not tx_done:
+        pass
+    print("Tx Done")
+
+    uart.irq(trigger=UART.RX_DONE, handler=rx_complete)
+    uart_rx_buf = bytearray(8)
+    uart.readinto(uart_rx_buf)
```
Comment (ramya-subramanyam) [inline tests/psoc6/hw_ext/uart.py:93]: Do we need to verify here the data read?

File: docs/psoc6/quickref.rst [hunk 2]
```diff
@@ -738,3 +738,68 @@ Constructor
...
+   - ``id`` takes any integer (This value is ignored for PSoC6 port)
+
+Methods
+^^^^^^^
+
+.. method:: UART.init(baudrate, bits, parity, stop, tx, rx, rts, cts, timeout,flow)
+
+    Initialize the UART with the given parameters.
+
+    Mandatory parameters:
```
Comment (ramya-subramanyam) [inline docs/psoc6/quickref.rst:764]: Some arguments are keyword arguments and some are not. Do we not document this here?

File: ports/psoc6/machine_uart.c [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+    { MP_ROM_QSTR(MP_QSTR_TX_DONE), MP_ROM_INT(CYHAL_UART_IRQ_TX_DONE) }, \
+    { MP_ROM_QSTR(MP_QSTR_RX_FULL), MP_ROM_INT(CYHAL_UART_IRQ_RX_FULL) }, \
+    { MP_ROM_QSTR(MP_QSTR_RX_DONE), MP_ROM_INT(CYHAL_UART_IRQ_RX_DONE) }, \
+
+machine_uart_obj_t *uart_obj[MAX_UART] = { NULL };
+
+static const char *_parity_name[] = {"None", "0", "1"};
+
+
+void uart_event_handler(void *handler_arg, cyhal_uart_event_t event) {
```
Comment (ramya-subramanyam) [inline ports/psoc6/machine_uart.c:66]: these 2 functions uart_event_handler and uart_irq_config could also be static functions right?

**Review by ramya-subramanyam — COMMENTED** [review]
Looks very good! just some minor comments I have added.

File: tests/psoc6/hw_ext/uart.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    uart_cts_pin = "P6_3"
+elif "CY8CPROTO-063-BLE" in machine:
+    uart_rx_pin = "P6_0"
+    uart_tx_pin = "P6_1"
+    uart_rts_pin = "P6_2"
+    uart_cts_pin = "P6_3"
+
+# 1. Construct instance
+##############################################
+uart = UART(1)
```
Comment (NikhitaR-IFX) [inline tests/psoc6/hw_ext/uart.py:23]: This confuses me. Why 1? It can be anything is it?

File: tests/psoc6/hw_ext/uart.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+    # read()
+    tx_data = b"abcdefg"
+    uart.write(tx_data)
+    time.sleep_ms(100)
+    rx_data = uart.read()
+    print("Tx is received by Rx(read()): ", rx_data == tx_data)
+
+    # readline(nbytes)
+    tx_data = "abcd\ne"
```
Comment (NikhitaR-IFX) [inline tests/psoc6/hw_ext/uart.py:50]: Is it typo after \n or intended?

**Review by NikhitaR-IFX — CHANGES_REQUESTED** [review]
Just small changes from my side. Rest looks good and if it works in HIL all green 👏👏

File: tests/psoc6/hw_ext/uart.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+from machine import UART
+import machine
+import binascii
+import time
+import os
+
+# Allocate pin based on board
+machine = os.uname().machine
+if "CY8CPROTO-062-4343W" in machine:
+    uart_rx_pin = "P6_0"
```
Comment (jaenrig-ifx) [inline tests/psoc6/hw_ext/uart.py:11]: Okay... it works.
But we need to find other pins as there are used by the SPI already. Can we find another compatible pins? Not used in these diagrams: https://github.com/Infineon/micropython/tree/ports-psoc6-ifx/tests/psoc6/hw_ext/img

File: ports/psoc6/machine_pin_phy.h [hunk 7]
```diff
@@ -10,7 +10,8 @@ typedef enum {
     PIN_PHY_FUNC_I2C = 0x04,
     PIN_PHY_FUNC_PWM = 0x08,
     PIN_PHY_FUNC_SPI = 0x10,
-    PIN_PHY_FUNC_I2S = 0x20
+    PIN_PHY_FUNC_I2S = 0x20,
+    PIN_PHY_FUNC_UART = 0x40
```
[the remark this answers is not in the document]
Comment (jaenrig-ifx) [reply ports/psoc6/machine_pin_phy.h:14]: Ah, I see the sdcard is now using 0x80. No need to change :)

File: tests/psoc6/hw_ext/uart.py (hunk 6 above)
Comment (IFX-Anusha) [reply tests/psoc6/hw_ext/uart.py]: Yes Updated with other pins.

**Review by jaenrig-ifx — APPROVED** [review]

**Review by ramya-subramanyam — APPROVED** [review]

**Review by NikhitaR-IFX — APPROVED** [review]

**Pull request merged**